README.md
# Homebridge plugin for [inio](https://inio-light.com) lights

## Still under development!

This plugin lets you control your inio lamps with Homekit.

currently working:
- brightness
- on/off
- set to "connected" mode (which follows the predetermined pattern by the app)

Exposes a lamp and a switch currently

## Installation

```
npm i homebridge-inio
```

or just add it in the homebridge UI


## Configuration

Example for one accesory
```
{
    accesory: "inioLight",
    name: "<same name>",
    url: "http://<ip of lamp>",
}
```

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/reckter"><img src="https://avatars.githubusercontent.com/u/1771450?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Hannes Güdelhöfer</b></sub></a><br /><a href="https://github.com/reckter/homebridge-inio/commits?author=reckter" title="Code">💻</a> <a href="#ideas-reckter" title="Ideas, Planning, & Feedback">🤔</a> <a href="#infra-reckter" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="#maintenance-reckter" title="Maintenance">🚧</a> <a href="https://github.com/reckter/homebridge-inio/commits?author=reckter" title="Documentation">📖</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->
## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

CHANGELOG.md
# v0.7.1 (Sat Feb 18 2023)

#### 🐛 Bug Fix

- Set the mode to static on brighntess / temperature changes [#6](https://github.com/reckter/homebridge-inio/pull/6) ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.7.0 (Thu Feb 16 2023)

#### 🚀 Enhancement

- Add color Temperature [#4](https://github.com/reckter/homebridge-inio/pull/4) ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.6.5 (Wed Feb 15 2023)

#### 🐛 Bug Fix

- Fix switch not switching on correctly [#3](https://github.com/reckter/homebridge-inio/pull/3) ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.6.4 (Wed Feb 15 2023)

#### ⚠️ Pushed to `main`

- Go back to github token ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.6.0 (Thu Feb 16 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, Hannes Güdelhöfer ([@reckter](https://github.com/reckter)), for all your work!

#### 🚀 Enhancement

- Fix attribution [#2](https://github.com/reckter/homebridge-inio/pull/2) ([@reckter](https://github.com/reckter))
- :tada: [#1](https://github.com/reckter/homebridge-inio/pull/1) ([@reckter](https://github.com/reckter))

#### ⚠️ Pushed to `main`

- Initial commit ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.5.0 (Thu Feb 16 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, Hannes Güdelhöfer ([@reckter](https://github.com/reckter)), for all your work!

#### 🚀 Enhancement

- Fix attribution [#2](https://github.com/reckter/homebridge-inio/pull/2) ([@reckter](https://github.com/reckter))
- :tada: [#1](https://github.com/reckter/homebridge-inio/pull/1) ([@reckter](https://github.com/reckter))

#### ⚠️ Pushed to `main`

- Initial commit ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))

---

# v0.4.0 (Thu Feb 16 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, Hannes Güdelhöfer ([@reckter](https://github.com/reckter)), for all your work!

#### 🚀 Enhancement

- Fix attribution [#2](https://github.com/reckter/homebridge-inio/pull/2) ([@reckter](https://github.com/reckter))
- :tada: [#1](https://github.com/reckter/homebridge-inio/pull/1) ([@reckter](https://github.com/reckter))

#### ⚠️ Pushed to `main`

- Initial commit ([@reckter](https://github.com/reckter))

#### Authors: 1

- Hannes Güdelhöfer ([@reckter](https://github.com/reckter))


scripts/generate-estimation-map.ts
import axios from 'axios';

import fs from 'fs';


// Change if you want to use a different lamp
const LAMP_URL = "http://192.168.161.151"

type LampPowers = {
    cold: number,
    warm: number
}

type LampTemperature = {
    brightness: number,
    kelvin: number
}

async function getLampPower(): Promise<LampPowers> {
    let response = (await axios(LAMP_URL + "/api/app/pwm_duty_get")).data
    let settled = false
    while (!settled) {
        await sleep(10)
        const current = (await axios(LAMP_URL + "/api/app/pwm_duty_get")).data
        settled = current.cold === response.cold && current.warm === response.warm
        response = current

    }
    return response
}

async function setLampBrightness(to: LampTemperature): Promise<void> {
    await axios(LAMP_URL + "/api/app/light_color", {
        method: 'POST',
        data: to
    })
}

async function getPowerForBrightness(to: LampTemperature, sleepLong?: boolean): Promise<LampPowers> {
    await setLampBrightness(to)
    // await sleep(sleepLong? 2000 : 100)
    return await getLampPower()
}

async function loadResult(): Promise<Array<Item>> {
    const result: Array<Item> = JSON.parse(fs.readFileSync('data.json', "utf-8"))
    return result
}

async function loadResultPacked(): Promise<Array<Item>> {
    const packed: Array<Array<number>> = JSON.parse(fs.readFileSync('src/data-packed.json', "utf-8"))
    return packed.map(it => ({
        temperature: {
            brightness: it[0],
            kelvin: it[1]
        },
        power: {
            warm: it[2],
            cold: it[3]
        }
    }))
}

/**
 * Generates a raibow table of all possible brightness and kelvin combinations
 * saves a snapshot to data-packed.json every so often
 * Can continue from such a snapshot
 */
async function mapAll() {
    const result: Array<{ power: LampPowers, temperature: LampTemperature }> = await loadResultPacked()

    const seen = new Set<string>(
        result
            .map(it => it.temperature)
            .map(it => `${it.brightness},${it.kelvin}`)
    )

    for (let brightness = 0; brightness <= 100; brightness += 1) {
        for (let kelvin = 2700; kelvin <= 6500; kelvin += 10) {
            const temperature = {brightness, kelvin}
            const key = `${brightness},${kelvin}`
            if (seen.has(key)) continue

            const power = await getPowerForBrightness(temperature, kelvin == 2700)

            console.log(`${brightness},${kelvin} -> ${power.cold},${power.warm}`)

            result.push({power, temperature})
            seen.add(key)
        }
        void writeCompact(result)
    }
    await sleep(1000)
    await writeCompact(result)
    return result
}


function distance(a: LampPowers, b: LampPowers): number {
    const coldDistance = Math.abs(a.cold - b.cold)
    const warmDistance = Math.abs(a.warm - b.warm)
    return coldDistance * coldDistance + warmDistance * warmDistance
}

type Item = {
    power: LampPowers,
    temperature: LampTemperature
}

function getEstimate(power: LampPowers, map: Array<{ power: LampPowers, temperature: LampTemperature }>): Item {
    let best = map[0]
    let bestDistance = distance(power, best.power)

    map.forEach(it => {
        if(distance(it.power, power) < bestDistance) {
            best = it
            bestDistance = distance(it.power, power)
        }
    })

    return best
}

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms))
}

/**
 * Randomly sets a temprateru and brightness and compares the result to the estimate
 * @param result
 */
async function randomTest(result: Array<{ power: LampPowers, temperature: LampTemperature }>) {
    const brightness = Math.floor(Math.random() * 101)
    const kelvin = (Math.floor(Math.random() * (6500 - 2700)) + 2700)

    await setLampBrightness({brightness, kelvin})
    const power = await getLampPower()
    const estimate = getEstimate(power, result)
    const dist = Math.sqrt(distance(power, estimate.power))
    const distKelvin = estimate.temperature.kelvin - kelvin
    const distBrightness = estimate.temperature.brightness - brightness
    console.log(`set: ${brightness},${kelvin}, power: ${power.cold},${power.warm} \n\t-> power: ${estimate.power.cold}, ${estimate.power.warm} estimate: ${estimate.temperature.brightness},${estimate.temperature.kelvin} (${dist}; ${distBrightness}, ${distKelvin})`)
}

/**
 * runns randomTest 100 times
 */
async function test() {
    const result = await loadResultPacked()
    for (let i = 0; i < 100; i++) {
        await randomTest(result)
    }
}


async function writeCompact(result: Array<Item>){
    const packed = Object.values(result)
        .map(it => [it.temperature.brightness, it.temperature.kelvin, it.power.warm, it.power.cold] )
        .map(it => it.map(it => parseFloat(it.toFixed(5))))

    await fs.promises.writeFile('src/data-packed.json', JSON.stringify(packed))
}
/**
 * Compacts the
 */
async function dedupe() {
    console.log("loading data")
    const data = await loadResultPacked()

    console.log("deduping")
    const map: Record<string, Item> = {}

   data.forEach((it) => {
        map[`${it.temperature.brightness},${it.temperature.kelvin}`] =it
    })

    console.log("packing")
    const packed = Object.values(map)
        .map(it => [it.temperature.brightness, it.temperature.kelvin, it.power.warm, it.power.cold] )
        .map(it => it.map(it => parseFloat(it.toFixed(5))))

    console.log("saving")
    fs.writeFileSync('src/data-packed.json', JSON.stringify(packed))
}

mapAll().then(dedupe)

src/accessory.ts
import type {
  AccessoryConfig,
  AccessoryPlugin,
  API,
  CharacteristicValue,
  Logging,
  Service,
} from 'homebridge';
import axios from 'axios';

type Config = AccessoryConfig & {
    url: string;
};

type LampPowers = {
    cold: number;
    warm: number;
};

type LampTemperature = {
    brightness: number;
    kelvin: number;
};

type EstimationItem = {
    power: LampPowers;
    temperature: LampTemperature;
};

type Mode = 'OFF' | 'CONNECTED' | 'STATIC';
type Addon = 'NONE' | 'BOOST' | 'RELAX' | 'SCENARIO';

type Status = {
    mode: Mode;
    addon: Addon;
};

import packedArray from './data-packed.json';

const estimationMap: Array<EstimationItem> = packedArray.map((it) => ({
  temperature: {
    brightness: it[0],
    kelvin: it[1],
  },
  power: {
    warm: it[2],
    cold: it[3],
  },
}));

export class InioAccessoryPlugin implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly config: Config;
  private api: API;

  private readonly lightBulb: Service;
  private readonly informationService: Service;
  private readonly modeSwitch: Service;

  constructor(log: Logging, config: AccessoryConfig, api: API) {
    this.log = log;
    this.config = config as Config;
    this.api = api;

    this.log.debug('Loading InioAccessoryPlugin');

    this.informationService = new this.api.hap.Service.AccessoryInformation()
      .setCharacteristic(this.api.hap.Characteristic.Manufacturer, 'Inio')
      .setCharacteristic(this.api.hap.Characteristic.Model, 'Inio Light');

    // expose a switch to turn on/off the "connected" mode
    this.modeSwitch = new this.api.hap.Service.Switch('Inio Light Mode');

    // this.modeSwitch.setCharacteristic(this.api.hap.Characteristic.Name, 'Inio Light Mode');
    // this.modeSwitch.setCharacteristic(this.api.hap.Characteristic.Model, 'Inio Light Mode');

    this.modeSwitch.getCharacteristic(this.api.hap.Characteristic.On)
    // .setProps({ description: "Turn on/off the 'connected' mode"})
      .onGet(this.handleModeGet.bind(this))
      .onSet(this.handleModeSet.bind(this));


    // expose a lightbulb with brightness
    this.lightBulb = new this.api.hap.Service.Lightbulb('Inio Light');

    this.lightBulb.getCharacteristic(this.api.hap.Characteristic.On)
      .onGet(this.handleOnGet.bind(this))
      .onSet(this.handleOnSet.bind(this));

    this.lightBulb.getCharacteristic(this.api.hap.Characteristic.Brightness)
      .onGet(this.handleBrightnesGet.bind(this))
      .onSet(this.handleBrightnesSet.bind(this));

    this.lightBulb.getCharacteristic(this.api.hap.Characteristic.ColorTemperature)
      .setProps({minValue: 153, maxValue: 370})
      .onGet(this.handleColorTemperatureGet.bind(this))
      .onSet(this.handleColorTemperatureSet.bind(this));
  }

  apiUrl(suffix: string) {
    return this.config.url.replace(/\/$/, '') + suffix;
  }


  async pushUpdate() {
    // this.lightBulb.updateCharacteristic(this.api.hap.Characteristic.On, await this.handleOnGet());
    // this.lightBulb.updateCharacteristic(this.api.hap.Characteristic.Brightness, await this.handleBrightnesGet());
    // this.lightBulb.updateCharacteristic(this.api.hap.Characteristic.ColorTemperature, await this.handleColorTemperatureGet());
    //
    // this.modeSwitch.updateCharacteristic(this.api.hap.Characteristic.On, await this.handleModeGet());
  }

  async getStatus(): Promise<Status> {
    const response = await axios(this.apiUrl('/api/app/status'));
    return response.data;
  }

  async setToMode(mode: Mode) {
    const status = await this.getStatus();
    if (status.mode === mode) {
      return;
    }

    switch (mode) {
      case 'OFF':
        await axios(this.apiUrl('/api/interface/btn_short_manual'), {method: 'PUT'});
        break;
      case 'CONNECTED':
        if (status.mode === 'OFF') {
          await axios(this.apiUrl('/api/interface/btn_short_manual'), {method: 'PUT'});
        }
        await axios(this.apiUrl('/api/interface/btn_short_connect'), {method: 'PUT'});
        break;
      case 'STATIC':
        if (status.mode === 'OFF') {
          await axios(this.apiUrl('/api/interface/btn_short_manual'), {method: 'PUT'});
        } else if (status.mode === 'CONNECTED') {
          await axios(this.apiUrl('/api/interface/btn_short_connect'), {method: 'PUT'});
        }
        break;
      default:
        this.log.error('Unknown mode: ' + mode);

    }
    await this.pushUpdate();
  }

  async handleModeGet(): Promise<CharacteristicValue> {
    const status = await this.getStatus();
    return status.mode === 'CONNECTED';
  }

  async handleModeSet(value: CharacteristicValue) {
    this.log.debug('handleModeSet', value);
    await this.setToMode(value ? 'CONNECTED' : 'OFF');
  }

  /**
     * euclidian distance
     */
  distance(a: LampPowers, b: LampPowers): number {
    const coldDistance = Math.abs(a.cold - b.cold);
    const warmDistance = Math.abs(a.warm - b.warm);
    return coldDistance * coldDistance + warmDistance * warmDistance;
  }

  /**
     * Gets the current lamp power
     * This is a bit tricky because the lamp takes some time to settle
     * So we wait until the values are stable
     */
  async getLampPower(): Promise<LampPowers> {
    let response = (await axios(this.apiUrl('/api/app/pwm_duty_get'))).data;
    let settled = false;
    this.log.debug('Getting lamp status');
    while (!settled) {
      this.log.debug('Waiting for lamp to settle');
      await sleep(100);
      const current = (await axios(this.apiUrl('/api/app/pwm_duty_get'))).data;
      settled = current.cold === response.cold && current.warm === response.warm;
      response = current;
    }
    this.log.debug('settled on ' + JSON.stringify(response));
    return response;
  }

  /**
     * Gets the estimated temperature based on the current power
     */
  getEstimate(power: LampPowers): LampTemperature {
    let best = estimationMap[0];
    let bestDistance = this.distance(power, best.power);

    estimationMap.forEach(it => {
      if (this.distance(it.power, power) < bestDistance) {
        best = it;
        bestDistance = this.distance(it.power, power);
      }
    });

    return best.temperature;
  }

  async getEstimatedTemperature(): Promise<LampTemperature> {
    const power = await this.getLampPower();
    return this.getEstimate(power);
  }

  async handleBrightnesGet(): Promise<CharacteristicValue> {
    const res = (await this.getEstimatedTemperature()).brightness;
    return res;
  }

  async handleBrightnesSet(value: CharacteristicValue) {
    try {
      await this.setToMode('STATIC');
      const temperature = await this.getEstimatedTemperature();
      await axios(this.apiUrl('/api/app/light_color'), {
        method: 'POST',
        data: {
          ...temperature,
          brightness: value,
        },
      });
      void this.pushUpdate();
    } catch (e) {
      this.log.error('got an error', e);
    }
  }

  async handleColorTemperatureGet(): Promise<CharacteristicValue> {
    const res = (await this.getEstimatedTemperature()).kelvin;
    // kelvin to mirek
    return 1_000_000 / res;
  }

  async handleColorTemperatureSet(value: CharacteristicValue) {
    try {
      await this.setToMode('STATIC');
      const temperature = await this.getEstimatedTemperature();
      // We need to clamp the value because the lamp does not support all values
      const kelvin = Math.floor(1_000_000 / (value as number));
      const clamped = Math.max(2700, Math.min(6500, kelvin));
      await axios(this.apiUrl('/api/app/light_color'), {
        method: 'POST',
        data: {
          ...temperature,
          kelvin: clamped,
        },
      });
      void this.pushUpdate();
    } catch (e) {
      this.log.error('got an error', e);
    }
  }

  async handleOnGet(): Promise<CharacteristicValue> {
    const status = await this.getStatus();
    return status.mode !== 'OFF';
  }

  async handleOnSet(value: CharacteristicValue) {
    try {
      await this.setToMode(value ? 'STATIC' : 'OFF');
      void this.pushUpdate();
    } catch (e) {
      this.log.error('got an error', e);
    }
  }

  getServices(): Service[] {
    return [
      this.informationService,
      this.lightBulb,
      this.modeSwitch,
    ];
  }
}


async function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}